src/darwin/Framework/CHIPTests/MTRCommissionableBrowserTests.m

```diff
@@ -0,0 +1,266 @@
+//
+//  MTRCommissionableBBrowserTests.m
+//  MTRCommissionableBBrowserTests
+/**
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#import <Matter/Matter.h>
+
+// system dependencies
+#import <XCTest/XCTest.h>
+
+#import "MTRTestKeys.h"
+#import "MTRTestOTAProvider.h"
+#import "MTRTestStorage.h"
+
+static const uint16_t kLocalPort = 5541;
+static const uint16_t kTestVendorId = 0xFFF1u;
+static const uint16_t kTestProductId = 0x8001u;
+static const uint16_t kTestDiscriminator = 3840u;
+static const uint16_t kDiscoverDeviceTimeoutInSeconds = 10;
+
+// Singleton controller we use.
+static MTRDeviceController * sController = nil;
+
+// Keys we can use to restart the controller.
+static MTRTestKeys * sTestKeys = nil;
+
+static BOOL sStackInitRan = NO;
+static BOOL sNeedsStackShutdown = YES;
+
+@interface DeviceScannerDelegate : NSObject <MTRCommissionableBrowserDelegate>
+@property (nonatomic) XCTestExpectation * expectation;
+
+- (instancetype)initWithExpectation:(XCTestExpectation *)expectation;
+- (void)didDiscoverCommissionable:(MTRDeviceController *)controller device:(MTRCommissionableBrowserResult *)device;
+- (void)commissionableUnavailable:(MTRDeviceController *)controller device:(MTRCommissionableBrowserResult *)device;
+@end
+
+@implementation DeviceScannerDelegate
+- (instancetype)initWithExpectation:(XCTestExpectation *)expectation
+{
+    if (!(self = [super init])) {
+        return nil;
+    }
+
+    _expectation = expectation;
+    return self;
+}
+
+- (void)didDiscoverCommissionable:(MTRDeviceController *)controller device:(MTRCommissionableBrowserResult *)device
+{
+    [self.expectation fulfill];
+
+    __auto_type instanceName = device.instanceName;
+    __auto_type vendorId = device.vendorID;
+    __auto_type productId = device.productID;
+    __auto_type discriminator = device.discriminator;
+    __auto_type commissioningMode = device.commissioningMode;
+
+    XCTAssertEqual([instanceName length], 16); // The  instance name is random, so just ensure the len is right.
+    XCTAssertEqualObjects(vendorId, @(kTestVendorId));
+    XCTAssertEqualObjects(productId, @(kTestProductId));
+    XCTAssertEqualObjects(discriminator, @(kTestDiscriminator));
```

Comment:
How can we assert that?  I would expect there to be three commissionables with three different discriminators around when this test runs.

More to the point, should we not hold off on the expectation fulfill until we have seen all three?

Reply:
I missed that we need 3 advertisements....